File: Data/Category.cs
using System.Collections.Generic;
using BooksStoreSPA.Data.DTO;

namespace BooksStoreSPA.Data
{
    public class Category : CategoryDTO
    {
        //для сортировки по имени родительской и текущей категории
        public string DisplayedName { get; set; }
        public Category ParentCategory { get; set; }
        public List<Category> ChildrenCategories { get; set; }
        public List<Book> Books { get; set; }
    }
}

File: ClientApp/src/app/data/category.ts
import { Book } from './book';
import { CategoryDTO } from './DTO/categoryDTO';

export class Category extends CategoryDTO {
    constructor(
        public displayedName?: string,
        public parentCategory?: Category,
        public childrenCategories?: Array<Category>,
        public books?: Array<Book>) {

        super();
    }
}
